source/common/common/shared_hash_map.h

```diff
@@ -0,0 +1,373 @@
+      uint32_t cell_index = control_->free_cell_index;
+      Cell& cell = getCell(cell_index);
+      control_->free_cell_index = cell.next_cell;
+      cell.next_cell = slots_[slot];
+      slots_[slot] = cell_index;
+      cell.key_size = key.size();
+      memcpy(cell.key, key.data(), key.size());
+      value = &cell.value;
+      ++control_->size;
+    }
+    unlock();
+    return value;
+  }
+
+  /**
+   * Removes the specified key from the map, returning bool if the key
+   * was found.
+   * @param key the key to remove
+   */
+  bool remove(absl::string_view key) {
+    if (key.size() > options_.max_key_size) {
+      return false;
+    }
+
+    bool found = false;
+    lock();
+    uint32_t slot = HashUtil::xxHash64(key) % options_.num_slots;
+    uint32_t* next = nullptr;
+    for (uint32_t* cptr = &slots_[slot]; *cptr != Sentinal; cptr = next) {
+      Cell& cell = getCell(*cptr);
+      next = &cell.next_cell;
+      if (cell.getKey() == key) {
+        control_->free_cell_index = *cptr;
+        --control_->size;
+        *cptr = *next; // Splices current cell out of slot-chain.
+        *next = control_->free_cell_index;
+        found = true;
+        break;
+      }
+    }
+    unlock();
+    return found;
+  }
+
+  /** Returns the number of key/values stored in the map. */
+  size_t size() const { return control_->size; }
+
+  /**
+   * Gets the value associated with a key, returning null if the value was not found.
+   * @param key
+   */
+  Value* get(absl::string_view key) {
+    if (key.size() > options_.max_key_size) {
+      return nullptr;
+    }
+
+    lock();
+    Value* value = getLockHeld(key);
+    unlock();
+    return value;
+  }
+
+private:
+  /**
+   * Represents control-values for the hash-table, including a mutex, which
+   * must gate all access to the internals.
+   */
+  struct Control {
+    std::string toString() const {
+      return fmt::format("{} size={} free_cell_index={}", options.toString(), size,
+                         free_cell_index);
+    }
+
+    pthread_mutex_t mutex;
+    SharedHashMapOptions options; // Options established at map construction time.
+    uint32_t size;                // Number of values currently stored.
+    uint32_t free_cell_index;     // Offset of first free cell.
+  };
+
+  /**
+   * Represents a value-cell, which is stored in a linked-list from each slot.
+   */
+  struct Cell {
+    /** Returns the key as a string_view. */
+    absl::string_view getKey() const { return absl::string_view(key, key_size); }
+
+    Value value;        // Templated value field.
+    uint32_t next_cell; // OFfset of next cell in map->cells_, terminated with Sentinal.
+    uint8_t key_size;   // size of key in bytes, or 0 if unused.
+    char key[];
+  };
+
+  // It seems like this is an obvious constexpr, but it won't compile as one.
+  static size_t alignment() {
+    return std::max(alignof(Cell), std::max(alignof(uint32_t), alignof(Control)));
```

Comment:
Should alignof(Value) be included here?

Reply:
No,I don't think so, because it's part of Cell.  I will note, however, that the CI seems to catch alignment errors that don't show up in my development environment, for some reason.

Reply:
Got it; that seems correct.